ksim/task/amp.py

```diff
@@ -357,6 +361,30 @@ def get_disc_losses(
         sim_disc_loss = jnp.mean((sim_disc_logits + 1) ** 2)
         return real_disc_loss, sim_disc_loss
 
+    def _grad_penalty(
+        self,
+        disc_model: PyTree,
+        real_motions: PyTree,  # (B, T, ...)
+        rng: PRNGKeyArray,
+    ) -> Array:
+        """R1 gradient penalty  (w.r.t. the *input* features)."""
+
+        def disc_on_motion(motion: PyTree, key: PRNGKeyArray) -> Array:
+            return self.call_discriminator(disc_model, motion, key).sum()
+
+        grad_fn = jax.grad(disc_on_motion, argnums=0)
```

Comment:
okay, this part really confused me. `real_motions` is not parametrized, so applying gradients to it doesn't make any sense - you're basically trying to "update" the real data, which is not what you want to do.

i have never seen the trick in the paper, but it looks like the idea is to minimize the discriminator gradients on the real samples. it should be something like this

```suggestion
        def disc_on_motion(disc_model: PyTree, motion: PyTree, key: PRNGKeyArray) -> Array:
            return self.call_discriminator(disc_model, motion, key).sum()

        grad_fn = jax.grad(disc_on_motion, argnums=0)
```

then you need to change how they're applied below as well